File: src/ts/ssh-keys/pkcs1KeyFormatter.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import {
	SshAlgorithms,
	KeyPair,
	Rsa,
	RsaParameters,
	BigInt,
	DerReader,
	DerWriter,
	DerType,
} from '@microsoft/dev-tunnels-ssh';
import { KeyFormatter, getKeyEncryptionAlgorithm } from './keyFormatter';
import { KeyData } from './keyData';

/** Provides import/export of the PKCS#1 key format. */
export class Pkcs1KeyFormatter implements KeyFormatter {
	private static readonly publicKeyType = 'RSA PUBLIC KEY';
	private static readonly privateKeyType = 'RSA PRIVATE KEY';

	public async import(keyData: KeyData): Promise<KeyPair | null> {
		if (!keyData) throw new TypeError('KeyData object expected.');

		if (!keyData.keyType) {
			// Automatically determine public or private by reading the first few bytes.
			try {
				const reader = new DerReader(keyData.data);
				reader.readInteger();
				reader.readInteger();
				keyData.keyType =
					reader.available > 0
						? Pkcs1KeyFormatter.privateKeyType
						: Pkcs1KeyFormatter.publicKeyType;
			} catch (e) {
				return null;
			}
		}

		let parameters: RsaParameters | null = null;
		if (keyData.keyType === Pkcs1KeyFormatter.publicKeyType) {
			parameters = Pkcs1KeyFormatter.parseRsaPublic(keyData.data);
		} else if (keyData.keyType === Pkcs1KeyFormatter.privateKeyType) {
			parameters = Pkcs1KeyFormatter.parseRsaPrivate(keyData.data);
		}

		if (parameters) {
			const keyPair = SshAlgorithms.publicKey.rsaWithSha512!.createKeyPair();
			await keyPair.importParameters(parameters);
			return keyPair;
		}

		return null;
	}

	public async export(keyPair: KeyPair, includePrivate: boolean): Promise<KeyData> {
		if (!keyPair) throw new TypeError('KeyPair object expected.');

		if (keyPair instanceof Rsa.KeyPair) {
			if (!keyPair.hasPublicKey) {
				throw new Error('KeyPair object does not have a public key.');
			} else if (includePrivate && !keyPair.hasPrivateKey) {
				throw new Error('KeyPair object does not have a private key.');
			}

			let keyData = new KeyData();
			const parameters = await keyPair.exportParameters();
			if (includePrivate) {
				keyData.keyType = Pkcs1KeyFormatter.privateKeyType;
				keyData.data = Pkcs1KeyFormatter.formatRsaPrivate(parameters);
			} else {
				keyData.keyType = Pkcs1KeyFormatter.publicKeyType;
				keyData.data = Pkcs1KeyFormatter.formatRsaPublic(parameters);
			}

			return keyData;
		} else {
			throw new Error('KeyPair class not supported.');
		}
	}

	public async decrypt(keyData: KeyData, passphrase: string | null): Promise<KeyData | null> {
		if (!keyData) throw new TypeError('KeyData object expected.');

		if (
			keyData.keyType === Pkcs1KeyFormatter.publicKeyType ||
			(!keyData.keyType && !passphrase)
		) {
			return keyData;
		} else if (keyData.keyType === Pkcs1KeyFormatter.privateKeyType || !keyData.keyType) {
			const procTypeHeader = keyData.headers.get('Proc-Type');
			if (procTypeHeader === '4,ENCRYPTED') {
				if (!passphrase) {
					throw new Error('A passphrase is required to decrypt the key.');
				}

				return Pkcs1KeyFormatter.decryptPrivate(keyData, passphrase);
			} else {
				return keyData;
			}
		}

		return null;
	}

	public async encrypt(keyData: KeyData, passphrase: string): Promise<KeyData> {
		if (!keyData) throw new TypeError('KeyData object expected.');

		if (keyData.keyType === Pkcs1KeyFormatter.publicKeyType) {
			throw new Error('Public key cannot be encrypted.');
		} else if (keyData.keyType === Pkcs1KeyFormatter.privateKeyType) {
			throw new Error(
				'PKCS#1 export with passphrase is not supported because the format uses ' +
					'a weak key derivation algorithm. Use PKCS#8 to export a ' +
					'passphrase-protected private key.',
			);
		} else {
			throw new Error(`Unsupported key type: ${keyData.keyType}`);
		}
	}

	private static formatRsaPublic(rsa: RsaParameters): Buffer {
		const writer = new DerWriter(Buffer.alloc(1024));
		writer.writeInteger(rsa.modulus);
		writer.writeInteger(rsa.exponent);
		return writer.toBuffer();
	}

	private static formatRsaPrivate(rsa: RsaParameters): Buffer {
		if (!(rsa.d && rsa.p && rsa.q && rsa.dp && rsa.dq && rsa.qi)) {
			throw new Error('Missing private key parameters.');
		}

		const writer = new DerWriter(Buffer.alloc(2048));
		writer.writeInteger(BigInt.fromInt32(0));
		writer.writeInteger(rsa.modulus);
		writer.writeInteger(rsa.exponent);
		writer.writeInteger(rsa.d);
		writer.writeInteger(rsa.p);
		writer.writeInteger(rsa.q);
		writer.writeInteger(rsa.dp);
		writer.writeInteger(rsa.dq);
		writer.writeInteger(rsa.qi);
		return writer.toBuffer();
	}

	private static parseRsaPublic(keyBytes: Buffer): RsaParameters {
		const reader = new DerReader(keyBytes);
		const modulus = reader.readInteger();
		const exponent = reader.readInteger();
		return { modulus, exponent };
	}

	private static parseRsaPrivate(keyBytes: Buffer): RsaParameters {
		const reader = new DerReader(keyBytes);
		const version = reader.readInteger();
		const modulus = reader.readInteger();
		const exponent = reader.readInteger();
		const d = reader.readInteger();
		const p = reader.readInteger();
		const q = reader.readInteger();
		const dp = reader.readInteger();
		const dq = reader.readInteger();
		const qi = reader.readInteger();
		return { modulus, exponent, d, p, q, dp, dq, qi };
	}

	/* @internal */
	public static async decryptPrivate(keyData: KeyData, passphrase: string): Promise<KeyData> {
		const decryptionInfo = keyData.headers.get('DEK-Info');
		if (!decryptionInfo) {
			throw new Error('PKCS#1 decryption parameters not found.');
		}

		const decryptionInfoParts = decryptionInfo.split(',');
		const cipherName = decryptionInfoParts[0];
		const ivHex = decryptionInfoParts[1];
		const iv = Buffer.from(ivHex, 'hex');

		const encryption = getKeyEncryptionAlgorithm(cipherName);
		const key = Pkcs1KeyFormatter.deriveDecryptionKey(
			Buffer.from(passphrase, 'utf8'),
			iv,
			encryption.keyLength,
		);

		const decryptedKeyData = new KeyData(keyData.keyType);
		decryptedKeyData.headers = new Map<string, string>(keyData.headers);
		decryptedKeyData.headers.delete('Proc-Type');
		decryptedKeyData.headers.delete('DEK-Info');

		const decipher = await encryption.createCipher(false, key, iv);
		decryptedKeyData.data = await decipher.transform(keyData.data);

		// The first part of the key should be a DER sequence header.
		if (decryptedKeyData.data[0] !== (DerType.Constructed | DerType.Sequence)) {
			throw new Error('Key decryption failed - incorrect passphrase.');
		}

		return decryptedKeyData;
	}

	private static deriveDecryptionKey(
		passphraseBytes: Buffer,
		iv: Buffer,
		keyLength: number,
	): Buffer {
		const useWebCrypto = !!(typeof crypto === 'object' && crypto.subtle);
		if (useWebCrypto) {
			// Web crypto does not provide an MD5 implementation. An external lib could be used here,
			// but it's not worth it to support an insecure encryption format. Use PKCS#8 instead.
			throw new Error('PKCS#1 decryption not implemented.');
		}

		const nodeCrypto = require('crypto');
		const PKCS5_SALT_LEN = 8;
		const salt = iv.slice(0, PKCS5_SALT_LEN);

		let key = Buffer.alloc(0);
		while (key.length < keyLength) {
			// MD5 is an unsafe hash algorithm, but it is used only for decrypting (importing) keys,
			// not encrypting.
			const md5 = nodeCrypto.createHash('md5');
			md5.update(key);
			md5.update(passphraseBytes);
			md5.update(salt);
			key = Buffer.concat([key, md5.digest()]);
		}

		key = key.slice(0, keyLength);
		return key;
	}
}

File: src/cs/Ssh.Keys/Pkcs1KeyFormatter.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using Microsoft.DevTunnels.Ssh.Algorithms;
using Microsoft.DevTunnels.Ssh.IO;

namespace Microsoft.DevTunnels.Ssh.Keys;

/// <summary>
/// Provides import/export of the <see cref="KeyFormat.Pkcs1" /> key format.
/// </summary>
public class Pkcs1KeyFormatter : IKeyFormatter
{
	private const string PublicKeyType = "RSA PUBLIC KEY";
	private const string PrivateKeyType = "RSA PRIVATE KEY";

	/// <inheritdoc/>
	public IKeyPair? Import(KeyData keyData)
	{
		if (keyData == null) throw new ArgumentNullException(nameof(keyData));

		if (string.IsNullOrEmpty(keyData.KeyType))
		{
			// Automatically determine public or private by reading the first few bytes.
			try
			{
				var reader = new DerReader(keyData.Data);
				reader.ReadInteger();
				reader.ReadInteger();
				keyData.KeyType = reader.Available > 0 ? PrivateKeyType : PublicKeyType;
			}
			catch (Exception)
			{
				return null;
			}
		}

		RSAParameters? parameters = null;
		if (keyData.KeyType == PublicKeyType)
		{
			parameters = ImportRsaPublic(keyData.Data);
		}
		else if (keyData.KeyType == PrivateKeyType)
		{
			parameters = ImportRsaPrivate(keyData.Data);
		}

		if (parameters != null)
		{
			Rsa.KeyPair? keyPair = new Rsa.KeyPair();
			keyPair.ImportParameters(parameters.Value);
			return keyPair;
		}

		return null;
	}

	/// <inheritdoc/>
	public KeyData Export(IKeyPair keyPair, bool includePrivate)
	{
		if (keyPair == null) throw new ArgumentNullException(nameof(keyPair));

		var keyAlgorithm = keyPair.KeyAlgorithmName;
		if (keyAlgorithm == Rsa.KeyAlgorithmName)
		{
			if (includePrivate)
			{
				return ExportRsaPrivate((Rsa.KeyPair)keyPair);
			}
			else
			{
				return ExportRsaPublic((Rsa.KeyPair)keyPair);
			}
		}
		else
		{
			throw new NotSupportedException(
				$"PKCS#1 format does not support key algorithm: {keyAlgorithm}");
		}
	}

	/// <inheritdoc/>
	public KeyData? Decrypt(KeyData keyData, string? passphrase)
	{
		if (keyData == null) throw new ArgumentNullException(nameof(keyData));

		if (keyData.KeyType == PublicKeyType ||
			(string.IsNullOrEmpty(keyData.KeyType) && string.IsNullOrEmpty(passphrase)))
		{
			return keyData;
		}
		else if (keyData.KeyType == PrivateKeyType ||
			(string.IsNullOrEmpty(keyData.KeyType) && !string.IsNullOrEmpty(passphrase)))
		{
			if (keyData.Headers.TryGetValue("Proc-Type", out var procType) &&
				procType == "4,ENCRYPTED")
			{
				if (string.IsNullOrEmpty(passphrase))
				{
					throw new UnauthorizedAccessException(
						"A passphrase is required to decrypt the key.");
				}

				return DecryptPrivate(keyData, passphrase!);
			}
			else
			{
				return keyData;
			}
		}

		return null;
	}

	/// <inheritdoc/>
	public KeyData Encrypt(KeyData keyData, string passphrase)
	{
		if (keyData == null) throw new ArgumentNullException(nameof(keyData));

		if (keyData.KeyType == PublicKeyType)
		{
			throw new ArgumentException("Public key cannot be encrypted.");
		}
		else if (keyData.KeyType == PrivateKeyType)
		{
			throw new NotSupportedException(
				"PKCS#1 export with passphrase is not supported because the format uses " +
				"a weak key derivation algorithm. Use a different format to export a " +
				"passphrase-protected private key.");
		}
		else
		{
			throw new InvalidOperationException($"Unexpected key type: {keyData.KeyType}");
		}
	}

	private static RSAParameters ImportRsaPublic(Buffer data)
	{
		var reader = new DerReader(data);
		return new RSAParameters
		{
			Modulus = reader.ReadInteger().ToByteArray(unsigned: true),
			Exponent = reader.ReadInteger().ToByteArray(unsigned: true),
		};
	}

	private static RSAParameters ImportRsaPrivate(Buffer data)
	{
		var reader = new DerReader(data);
		reader.ReadInteger(); // skip version

		return new RSAParameters
		{
			Modulus = reader.ReadInteger().ToByteArray(unsigned: true),
			Exponent = reader.ReadInteger().ToByteArray(unsigned: true),
			D = reader.ReadInteger().ToByteArray(unsigned: true),
			P = reader.ReadInteger().ToByteArray(unsigned: true),
			Q = reader.ReadInteger().ToByteArray(unsigned: true),
			DP = reader.ReadInteger().ToByteArray(unsigned: true),
			DQ = reader.ReadInteger().ToByteArray(unsigned: true),
			InverseQ = reader.ReadInteger().ToByteArray(unsigned: true),
		};
	}

	private static KeyData ExportRsaPublic(Rsa.KeyPair keyPair)
	{
		var parameters = keyPair.ExportParameters(false);

		var writer = new DerWriter();
		writer.WriteInteger(BigInt.FromByteArray(parameters.Modulus!, unsigned: true));
		writer.WriteInteger(BigInt.FromByteArray(parameters.Exponent!, unsigned: true));

		var data = writer.ToArray();

		return new KeyData
		{
			KeyType = "RSA PUBLIC KEY",
			Data = data,
		};
	}

	private static KeyData ExportRsaPrivate(Rsa.KeyPair keyPair)
	{
		var parameters = keyPair.ExportParameters(true);

		var writer = new DerWriter();

		writer.WriteInteger(BigInt.FromInt32(0)); // version

		writer.WriteInteger(BigInt.FromByteArray(parameters.Modulus!, unsigned: true));
		writer.WriteInteger(BigInt.FromByteArray(parameters.Exponent!, unsigned: true));
		writer.WriteInteger(BigInt.FromByteArray(parameters.D!, unsigned: true));
		writer.WriteInteger(BigInt.FromByteArray(parameters.P!, unsigned: true));
		writer.WriteInteger(BigInt.FromByteArray(parameters.Q!, unsigned: true));
		writer.WriteInteger(BigInt.FromByteArray(parameters.DP!, unsigned: true));
		writer.WriteInteger(BigInt.FromByteArray(parameters.DQ!, unsigned: true));
		writer.WriteInteger(BigInt.FromByteArray(parameters.InverseQ!, unsigned: true));

		var data = writer.ToArray();

		return new KeyData
		{
			KeyType = "RSA PRIVATE KEY",
			Data = data,
		};
	}

	internal static KeyData DecryptPrivate(KeyData keyData, string passphrase)
	{
		if (!keyData.Headers.TryGetValue("DEK-Info", out var decryptionInfo))
		{
			throw new NotSupportedException("PKCS#1 decryption parameters not found.");
		}

		var decryptionInfoParts = decryptionInfo.Split(',');
		var cipherName = decryptionInfoParts[0];

		var ivHex = decryptionInfoParts[1];
		byte[] iv = new byte[ivHex.Length / 2];
		for (int i = 0; i < iv.Length; i++)
			iv[i] = Convert.ToByte(ivHex.Substring(i * 2, 2), 16);

		var encryption = KeyPair.GetKeyEncryptionAlgorithm(cipherName);

		var key = DeriveDecryptionKey(
			Encoding.UTF8.GetBytes(passphrase), iv, encryption.KeyLength);

		var decryptedKeyData = new KeyData
		{
			KeyType = keyData.KeyType,
			Headers = keyData.Headers,
			Data = new Buffer(keyData.Data.Length).Array,
		};

		decryptedKeyData.Headers.Remove("Proc-Type");
		decryptedKeyData.Headers.Remove("DEK-Info");

		using var decipher = encryption.CreateCipher(isEncryption: false, key, iv);
		decipher.Transform(keyData.Data, decryptedKeyData.Data);

		// The first part of the key should be a DER sequence header.
		if (decryptedKeyData.Data[0] != (byte)(DerType.Constructed | DerType.Sequence))
		{
			throw new UnauthorizedAccessException(
				"Key decryption failed - incorrect passphrase.");
		}

		return decryptedKeyData;
	}

	private static byte[] DeriveDecryptionKey(
		byte[] passphraseBytes, byte[] iv, int keyLength)
	{
		// Reference EVP_BytesToKey() from
		// https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/evp/evp_key.c#L74

		// Justification: MD5 is used only for decrypting (importing) keys, not encrypting.
#pragma warning disable CA5350 // Do Not Use Weak Cryptographic Algorithms
#pragma warning disable CA5351 // Do Not Use Broken Cryptographic Algorithms
		using HashAlgorithm hash = MD5.Create();
#pragma warning restore CA5350 // Do Not Use Weak Cryptographic Algorithms
#pragma warning restore CA5351 // Do Not Use Broken Cryptographic Algorithms

		const int PKCS5_SALT_LEN = 8;
		var salt = iv.Take(PKCS5_SALT_LEN);

		byte[] key = Array.Empty<byte>();
		while (key.Length < keyLength)
		{
			var digest = hash.ComputeHash(key.Concat(passphraseBytes).Concat(salt).ToArray());
			key = key.Concat(digest).ToArray();
		}

		key = key.Take(keyLength).ToArray();
		return key;
	}
}
